Bootstraps/AbstractSymfony.php

```diff
@@ -0,0 +1,224 @@
+
+/**
+ * Shared bootstrap logic for Symfony Kernels. Handles changes introduced in Symfony 3.3/4 and 4.0
+ */
+abstract class AbstractSymfony implements BootstrapInterface, HooksInterface, ApplicationEnvironmentAwareInterface
+{
+
+    /**
+     * @var string|null The application environment
+     */
+    protected $appenv;
+
+    /**
+     * @var boolean
+     */
+    protected $debug;
+
+    /**
+     * Instantiate the bootstrap, storing the $appenv
+     *
+     * @param $appenv
+     * @param $debug
+     */
+    public function initialize($appenv, $debug)
+    {
+        $this->appenv = $appenv;
+        $this->debug = $debug;
+    }
+
+    /**
+     * Does some necessary preparation before each request.
+     *
+     * @param KernelInterface $app
+     */
+    public function preHandle($app)
+    {
+        //resets Kernels startTime, so Symfony can correctly calculate the execution time
+        Utils::hijackProperty($app, 'startTime', microtime(true));
+    }
+
+    /**
+     * Does some necessary clean up after each request.
+     *
+     * @param KernelInterface $app
+     */
+    public function postHandle($app)
+    {
+        $container = $app->getContainer();
+
+        //resets stopwatch, so it can correctly calculate the execution time
+        if ($container->has('debug.stopwatch')) {
+            $container->get('debug.stopwatch')->__construct();
+        }
+
+        //Symfony\Bundle\TwigBundle\Loader\FilesystemLoader
+        //->Twig_Loader_Filesystem
+        if ($container->has('twig.loader')) {
+            $twigLoader = $container->get('twig.loader');
+            Utils::bindAndCall(function() use ($twigLoader) {
+                foreach ($twigLoader->cache as $path) {
+                    ppm_register_file($path);
+                }
+            }, $twigLoader);
+        }
+
+        //reset all profiler stuff currently supported
+        if ($container->has('profiler')) {
+            $profiler = $container->get('profiler');
+
+            // since Symfony does not reset Profiler::disable() calls after each request, we need to do it,
+            // so the profiler bar is visible after the second request as well.
+            $profiler->enable();
+
+            //PropelLogger
+            if ($container->has('propel.logger')) {
+                $propelLogger = $container->get('propel.logger');
+                Utils::hijackProperty($propelLogger, 'queries', []);
+            }
+
+            //Doctrine
+            //Doctrine\Bundle\DoctrineBundle\DataCollector\DoctrineDataCollector
+            if ($profiler->has('db')) {
+                Utils::bindAndCall(function () {
+                    //$logger: \Doctrine\DBAL\Logging\DebugStack
+                    foreach ($this->loggers as $logger){
+                        Utils::hijackProperty($logger, 'queries', []);
+                    }
+                }, $profiler->get('db'), null, 'Symfony\Bridge\Doctrine\DataCollector\DoctrineDataCollector');
```

Comment:
Why not to use ::class?

Reply:
It's from the original Symfony bootstrap adapter. I believe it was written that way to handle cases where Doctrine is not loaded.